### PR TITLE
fossid-webapp: Delta scans

### PR DESCRIPTION
**Pull request by nnobelis**
When set, ORT will create delta scans. When only changes in a repository
need to be scanned, delta scans reuse the identifications of latest scan
on the same repository to reduce the amount of findings.
If deltaScans is set and no scan exist yet, an initial scan called
"origin" scan will be created.

Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 1]
```diff
@@ -162,6 +166,13 @@ class FossId(
         }
     }
 
+    /**
+     * The qualifier of a scan when delta scans are enabled.
+     */
+    internal enum class DeltaTag {
+        ORIGIN, DELTA
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:173]: Can you please document each enum member?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 2]
```diff
@@ -47,26 +49,33 @@ class FossIdNamingProvider(
...
+            "#projectName" to projectName
+        )
+        replaceNamingConventionVariables(namingProjectPattern, builtins, namingConventionVariables)
     } ?: projectName
 
-    fun createScanCode(projectName: String): String {
-        val pattern = namingScanPattern ?: "${projectName}_#currentTimestamp"
-        return replaceNamingConventionVariables(pattern, projectName, namingConventionVariables)
+    fun createScanCode(projectName: String, deltaTag: FossId.DeltaTag? = null): String {
+        val defaultPattern = "${projectName}_#currentTimestamp${if (deltaTag == null) "" else "_#deltaTag"}"
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:59]: BTW, it just occurred to me: Why is `${projectName}` already inserted here? Shouldn't it be `#projectName` to let `replaceNamingConventionVariables()` do the replacement?

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt]: Oh yes, well spotted.

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 3]
```diff
@@ -47,26 +49,33 @@ class FossIdNamingProvider(
...
+            "#projectName" to projectName
+        )
+        replaceNamingConventionVariables(namingProjectPattern, builtins, namingConventionVariables)
     } ?: projectName
 
-    fun createScanCode(projectName: String): String {
-        val pattern = namingScanPattern ?: "${projectName}_#currentTimestamp"
-        return replaceNamingConventionVariables(pattern, projectName, namingConventionVariables)
+    fun createScanCode(projectName: String, deltaTag: FossId.DeltaTag? = null): String {
+        val defaultPattern = "#projectName_#currentTimestamp${if (deltaTag == null) "" else "_#deltaTag"}"
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:59]: Thanks for the pattern fix, but ideally it would go to a separate commit (the first) in this PR, as it has been wrong before, and it's an independent change.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 4]
```diff
@@ -319,35 +339,15 @@ class FossId(
...
-                    scanCode
+                val scanCode = if (deltaScans) {
+                    checkAndCreateDeltaScan(scans, revision, url, projectCode, projectName)
                 } else {
-                    log.info { "Scan ${existingScan.code} found for $url and revision $revision." }
-
-                    requireNotNull(existingScan.code) {
-                        "FossId returned a null scancode for an existing scan"
-                    }
+                    checkAndCreateScan(scans, revision, url, projectCode, projectName)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:345]: For a more pleasant review experience, could also the factoring-out of this function go to a separate (non-functional-change) commit? That allows a reviewer to only skim over commit that do not change / add logic, and to concentrate on the real changes.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 5]
```diff
@@ -394,6 +416,64 @@ class FossId(
                 "FossId returned a null scancode for an existing scan"
             }
         }
+        if (waitForResult) checkScan(scanCode)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:419]: Why does this check now happen *after* getting the raw result?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 6]
```diff
@@ -35,8 +35,10 @@ import org.ossreviewtoolkit.utils.log
  * Available builtin variables:
  * * **projectName**: The name of the project (i.e. the part of the URL before .git).
  * * **currentTimestamp**: The current time.
+ * * **deltaTag** (scan code only): If delta scans is enabled, this qualify the scan as an *origin* scan or a *delta*
+ * scan.
  */
-class FossIdNamingProvider(
+internal class FossIdNamingProvider(
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:41]: This lowering of the visibility is also unrelated to this change, strictly spoken, or?

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt [hunk 7]
```diff
@@ -47,26 +49,33 @@ class FossIdNamingProvider(
...
-        val pattern = namingScanPattern ?: "#projectName_#currentTimestamp"
-        return replaceNamingConventionVariables(pattern, projectName, namingConventionVariables)
+    fun createScanCode(projectName: String, deltaTag: FossId.DeltaTag? = null): String {
+        val defaultPattern = "#projectName_#currentTimestamp${if (deltaTag == null) "" else "_#deltaTag"}"
+        val pattern = namingScanPattern ?: defaultPattern
+        val builtins = mutableMapOf("#projectName" to projectName)
+        deltaTag?.let {
+            builtins += "#deltaTag" to deltaTag.name.lowercase()
+        }
+        return replaceNamingConventionVariables(pattern, builtins, namingConventionVariables)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:65]: How about checking for `deltaTag != null` only once (and inlining `pattern`), like

        var defaultPattern = "#projectName_#currentTimestamp"
        val builtins = mutableMapOf("#projectName" to projectName)

        if (deltaTag != null) {
            defaultPattern += "_#deltaTag"
            builtins += "#deltaTag" to deltaTag.name.lowercase()
        }

        return replaceNamingConventionVariables(namingScanPattern ?: defaultPattern, builtins, namingConventionVariables)

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 5 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:419]: Heu, no it still happen before: the call is in `checkAndCreateScan` and the latter is called before `getRawResults`.

File: scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt (hunk 6 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossIdNamingProvider.kt:41]: No it is required because `createScanCode` takes a `DeltaTag` in parameters which is `internal` and leads to:
`'public' function exposes its 'internal' parameter type DeltaTag`

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 5 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:419]: What I meant is: Is it intended that the original `checkScan` call from line 328 is removed?

PS: What does "Heu" mean? 😀

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:419]: Yes it is intended because the call moved to `checkAndCreateScan` and `checkAndCreateDeltaScan`.
Heu = Hrm in french :)

**Review by sschuberth — APPROVED** [review]
